### PR TITLE
CLI

### PR DESCRIPTION
**Pull request by leonardoalt**
based on ZoKrates' CLI

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -8,3 +8,4 @@ edition = "2021"
 [dependencies]
 yultsur = { git = "https://github.com/leonardoalt/yultsur", branch = "main" }
 tempfile = "3"
+clap = "2.26.2"
```
Comment (chriseth) [inline Cargo.toml:11]: What is the review policy for adding external dependencies? Yolo?

Comment (leonardoalt) [reply Cargo.toml:11]: It's curated, I copied from ZoKrates.

Comment (leonardoalt) [reply Cargo.toml:11]: Ask @Schaeff what his policy was

File: src/main.rs [hunk 2]
```diff
@@ -9,8 +10,45 @@ use yultsur::dialect::EVMDialect;
...
+
+    match matches.subcommand() {
+        ("symbolic", Some(sub_matches)) => symbolic_revert(sub_matches),
+        _ => unreachable!(),
+    }
+}
+
+fn symbolic() -> App<'static, 'static> {
+    SubCommand::with_name("symbolic")
+        .about("Symbolically execute Yul programs checking for reverts")
```
Comment (chriseth) [inline src/main.rs:38]:
```suggestion
        .about("Symbolically execute Yul programs checking for revert reachability")
```

File: src/main.rs [hunk 3]
```diff
@@ -19,14 +57,16 @@ fn main() {
     let mut ast = yul_parser::parse_block(&content);
     resolve::<EVMDialect>(&mut ast);
 
-    let query = yools::encoder::encode::<EVMInstructions>(&ast);
+    let query = yools::encoder::encode_revert_unreachable::<EVMInstructions>(&ast);
 
     match solver::query_smt(&query) {
         true => {
-            println!("SAT");
+            println!("Revert happens.");
```
Comment (chriseth) [inline src/main.rs:64]:
```suggestion
            println!("Revert is reachable.");
```

File: src/main.rs [hunk 4]
```diff
@@ -19,14 +57,16 @@ fn main() {
...
         }
         false => {
-            println!("UNSAT");
+            println!("Revert does not happen.");
```
Comment (chriseth) [inline src/main.rs:67]:
```suggestion
            println!("All reverts are unreachable.");
```

File: src/main.rs [hunk 5]
```diff
@@ -9,8 +10,45 @@ use yultsur::dialect::EVMDialect;
...
+        .subcommands(vec![symbolic()])
+        .get_matches();
+
+    match matches.subcommand() {
+        ("symbolic", Some(sub_matches)) => symbolic_revert(sub_matches),
+        _ => unreachable!(),
+    }
+}
+
+fn symbolic() -> App<'static, 'static> {
```
Comment (chriseth) [inline src/main.rs:36]: Maybe rename to indicate that this just returns the options?

**Comment by leonardoalt** [comment]
Comments fixed. Also changed clap's version to newest.

File: src/main.rs [hunk 6]
```diff
@@ -9,8 +10,45 @@ use yultsur::dialect::EVMDialect;
...
+                .value_name("FILE.yul")
+                .takes_value(true)
+                .required(true),
+        )
+}
+
+fn symbolic_revert(sub_matches: &ArgMatches) -> Result<(), String> {
+    let yul_file = PathBuf::from(sub_matches.value_of("input").unwrap());
 
     let mut file = File::open(yul_file).unwrap();
```
Comment (Schaeff) [inline src/main.rs:53]: This could be a single line:

```
let content = std::fs::read_to_string(yul_file).unwrap();
```

**Review by Schaeff — APPROVED** [review]
Looks good. `clap` is what everybody uses for CLI in Rust and has been since before I started writing Rust.

File: src/main.rs (hunk 6 above)
Comment (Schaeff) [reply src/main.rs:53]: oops, I meant line 53

**Review by chriseth — APPROVED** [review]

**Pull request merged**